test/runnable/test17899.d

```diff
@@ -0,0 +1,15 @@
+module test17899;
+
+int setme = 0;
+void delegate() bar1 = (){ setme = 1;};
+
+__gshared void delegate() bar2 = (){ setme = 2;};
+
+void main()
+{
+    assert(setme == 0);
+    bar1();
+    assert(setme == 1);
+    bar2();
+    assert(setme == 2);
+}
```

Comment:
Test failed.  The logged output:
/home/braddr/sandbox/at-client/pull-2994750-FreeBSD_32/dmd/generated/freebsd/release/32/dmd -conf= -m32 -Irunnable   -odgenerated/runnable -ofgenerated/runnable/test17899_0 runnable/test17899.d
generated/runnable/test17899_0


==============================
Test failed: caught signal: 11

Reply:
adding 
```d
    writeln(cast(void*)bar1.funcptr," ",cast(void*)bar1.ptr);
    writeln(cast(void*)bar2.funcptr," ",cast(void*)bar2.ptr);
```
yields 
```
null 10BD59174
null 10BD591B4
```
Looks like the `.funcptr` is being initialised with the `.ptr` and visa versa.
Any help much appreciated.

Reply:
yep, they are the wrong way round.
```d
void swap(T)(ref T t)
{
    auto tmp = t.ptr;
    t.ptr = cast(void*)t.funcptr;
    t.funcptr = cast(typeof(t.funcptr)) tmp;
}
void main()
{
    assert(setme  == 0);
    swap(bar1);
    bar1();
    assert(setme == 1);
    swap(bar2);
    bar2();
    assert(setme == 2);
}
```
works.

Reply:
Added padding so that `funcptr` should align with the runtime offset. Hopefully this works.